layout: Avoid flex item relayout sometimes when stretching

mrobinson opened the pull request:
This is the second flexbox caching change. It seeks to detect when a
relayout can be avoided in the case of a stretching flex item. This
heuristic can be combined, because currently we still do relayout
sometimes when we do not need to.

For instance currently we always relayout when a flex child is itself a
column flex. This only needs to happen when the grandchildren themselves
grow or shrink. That optimization is perhaps a lower priority as
`flex-grow: 0 / flex-shrink: 1` is the default behavior for flex.

Fixes #33567.

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because they do not change layout results and there is no way to measure performance with tests yet.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
layout: Avoid layout sometimes when stretching

This is the second flexbox caching change. It seeks to detect when a
relayout can be avoided in the case of a stretching flex item. This
heuristic can be combined, because currently we still do relayout
sometimes when we do not need to.

For instance currently we always relayout when a flex child is itself a
column flex. This only needs to happen when the grandchildren themselves
grow or shrink. That optimization is perhaps a lower priority as
`flex-grow: 0 / flex-shrink: 1` is the default behavior for flex.

Since this change means we more consistenly zero out the percentage part
of `calc` expressions when they have circular dependencies, this causes one
test to start failing (`/css/css-values/calc-min-height-block-1.html`).
This is related to w3c/csswg-drafts#10969, which is pending on further
discussion in the working group.

Co-authored-by: Oriol Brufau <obrufau@igalia.com>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson commented:
This fixes #33567 in the sense that the page is usable, but there's still plenty of optimization to do on layout.

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -1835,6 +1960,12 @@ impl FlexItem<'_> {
                 );
 
+                let has_child_which_depends_on_block_constraints = fragments.iter().any(|fragment| {
+                        fragment.base().map_or(false,|base|
```
```suggestion
                        fragment.base().is_some_and(|base|
```

Loirooriol reviewed (approved):
> `flex-grow: 1` is the default behavior for flex.

The initial values are `flex-grow: 0` and `flex-shrink: 1`.

> These changes do not require tests because they do not change layout results

Should explain `/css/css-values/calc-min-height-block-1.html`.
Or maybe first change `content_min_box_size()` in its own PR, so that we can then land this without behavior changes.

mrobinson commented:
> Should explain `/css/css-values/calc-min-height-block-1.html`. Or maybe first change `content_min_box_size()` in its own PR, so that we can then land this without behavior changes.

Sorry, I had fixed this in the commit message, but forgot to update the PR description.

mrobinson commented on components/layout_2020/style_ext.rs:
```diff
@@ -1042,6 +1020,25 @@ impl ComputedValuesExt for ComputedValues {
+        let has_percentage = |offset: LengthPercentageOrAuto<'_>| {
+            offset
+                .non_auto()
+                .map_or(false, |offset| offset.has_percentage())
```
Hrm. I agree with this change, but it's unusual that the Lint job didn't complain. I should be failing on new clippy warnings.

Loirooriol replied:
I recalled #33468 but it doesn't actually mention clippy so I guess clippy doesn't complain.

mrobinson commented:
> > Should explain `/css/css-values/calc-min-height-block-1.html`. Or maybe first change `content_min_box_size()` in its own PR, so that we can then land this without behavior changes.
> 
> Sorry, I had fixed this in the commit message, but forgot to update the PR description.

In the end I have removed this because you addressed it in #33988.
